File: problems/0008_string-to-integer-atoi/one-loop/Solution.swift
class Solution {
    func myAtoi(_ str: String) -> Int {
        if str.count == 0 {
            return 0
        }
        let chars = Array(str)
        var i = 0
        while i < str.count && chars[i] == " " {
            i += 1
        }
        var sign = 1
        if i < str.count {
            if chars[i] == "-" {
                i += 1
                sign = -1
            } else if chars[i] == "+" {
                i += 1
            }
        }
        var result = 0
        while i < str.count && "0" <= chars[i] && chars[i] <= "9" {
            result = result * 10 + sign * Int(chars[i].asciiValue! - 48)
            if result < INT_MIN {
                return INT_MIN
            } else if result > INT_MAX {
                return INT_MAX
            }
            i += 1
        }
        return result
    }
    let (INT_MIN, INT_MAX) = (Int(Int32.min), Int(Int32.max))
}

File: problems/0008_string-to-integer-atoi/one-loop/solution.ts
function myAtoi(str: string): number {
    if (str.length === 0)
        return 0
    let i = 0
    while (i < str.length && str[i] === ' ') ++i
    let sign = 1
    if (i < str.length)
        if (str[i] === '-') {
            ++i
            sign = -1
        } else if (str[i] === '+')
            ++i
    let result = 0
    for (; i < str.length && '0' <= str[i] && str[i] <= '9'; ++i) {
        result = result * 10 + sign * (str.charCodeAt(i) - CHAR_ZERO)
        if (result < INT_MIN)
            return INT_MIN
        else if (result > INT_MAX)
            return INT_MAX
    }
    return result
};

const INT_MAX = Math.pow(2, 31) - 1,
      INT_MIN = -Math.pow(2, 31),
      CHAR_ZERO = '0'.charCodeAt(0)
